apps/discord-bot/src/triggers/cta-post.ts

```diff
@@ -0,0 +1,280 @@
+        reason: 'Tracking CTA participation.',
+      });
+
+      const threadChannel = await thread;
+      threadChannel.send({
+        content: `Mark your completion of the CTA by reacting with ${finishedEmoji}.\n\n A chart will be posted in an hour, and updated every hour until the thread is archived, showing the number of completions per region.`,
+        flags: [MessageFlags.SuppressNotifications]
+      });
+      msg.react(finishedEmoji)
+
+      return thread
+    }
+
+    return undefined
+  }
+
+  // start a reaction collector on the given message
+  private async startCTAReactionCollector(msg: Message, thread: PublicThreadChannel): Promise<void> {
+    const filter = (reaction, user) => reaction.emoji.name === finishedEmoji && user.id != msg.client.user.id
+    const collector = msg.createReactionCollector({ filter, time: oneHour });
+    const roleReactions = await this.getFinishedReactions(msg)
+
+    collector.on('collect', async (reaction, user) => {
+      const member = msg.guild?.members.fetch(user.id)
+      if (member != undefined) {
+
+        console.log(`Got ${reaction.emoji.name} from ${(await member).displayName}`)
+
+        const memberRoles = (await member).roles.cache;
+        const role = memberRoles.filter(r => regionRoles.includes(r.name));
+        const memberRegionRole = role.first()?.name;
+
+        if (memberRegionRole != undefined) {
+          if (!roleReactions[memberRegionRole]) {
+            roleReactions[memberRegionRole] = [];
+          }
+          if (!roleReactions[memberRegionRole]?.includes(user.id)) {
+            roleReactions[memberRegionRole]?.push(user.id);
+            console.log(`Adding ${role.first()?.name} member, ${user.displayName}`);
+          }
+        }
+
+        this.sendChartToThread(msg, thread, roleReactions);
+      }
+    });
+
+    collector.on('end', collected => {
+      console.log(`Collected ${collected.size} reactions.`);
+      this.sendChartToThread(msg, thread, roleReactions);
+    });
+  }
+
+  private async deleteLastChart(tc: PublicThreadChannel): Promise<void> {
+    const threadMessages = await tc.messages.fetch({ limit: 1 });
+    const latest = threadMessages.first();
+    if (latest && latest.attachments.size > 0) {
+      await latest.delete();
+    }
+  }
+
+  private async sendChartToThread(msg: Message, thread: PublicThreadChannel, roleReactions: object): Promise<void> {
+    const pngBuffer = this.createChart(roleReactions).toBuffer('image/png');
+
+    writeFile('./misc/tmp/reaction_summary.png', pngBuffer, (err) => {
+      if (err) throw err;
+    });
+
+    (async () => {
+      const threadChannel = await thread;
+      if (!threadChannel.archived) {
+        await this.deleteLastChart(threadChannel);
+        await threadChannel.send({
+          files: ["./misc/tmp/reaction_summary.png"],
+          flags: [MessageFlags.SuppressNotifications]
+        });
+        this.startCTAReactionCollector(msg, thread);
+      }
+    })();
```

Comment:
I think there's potential for a race condition here. The call to `writeFile` is not being awaited, so the anonymous async function below isn't going to wait for it to finish. It could finish writing before it tries to send the image, but if it doesn't, the bot may send a corrupted PNG or throw an error about another process using the file.

I think a better approach would be to use the promise version of writeFile (https://nodejs.org/api/fs.html#promise-example) which can be awaited, then run the other parts (also I'm not sure why this is wrapped inside an anonymous function, can you explain?)

Reply:
Even better, avoid having to write the file to disk altogether by [converting the canvas to a PNG stream](https://www.npmjs.com/package/canvas#canvascreatepngstream) and using that with an [AttachmentBuilder](https://discord.js.org/docs/packages/discord.js/main/AttachmentBuilder:Class#constructor). The builder can be passed in the array on line 188.